File: gatedetection.py
import cv2
import numpy as np

def detect_gate(video_path, lower_red, upper_red):

    cap = cv2.VideoCapture(video_path)

    if not cap.isOpened():
        print("Error: Could not open video.")
        return

    while cap.isOpened():
        ret, frame = cap.read()

        if not ret:
            break

        hsv_frame = cv2.cvtColor(frame, cv2.COLOR_BGR2HSV)
        mask = cv2.inRange(hsv_frame, lower_red, upper_red)
        kernel = np.ones((5, 5), np.uint8)
        mask = cv2.dilate(mask, kernel, iterations=1)
        mask = cv2.erode(mask, kernel, iterations=1)
        
        edges = cv2.Canny(mask, threshold1=50, threshold2=150)
        
        lines = cv2.HoughLinesP(edges, 1, np.pi/180, threshold=50, minLineLength=50, maxLineGap=5)

        if lines is not None:
            for line in lines:
                x1, y1, x2, y2 = line[0]
                cv2.line(frame, (x1, y1), (x2, y2), (0, 255, 0), 2)
                
                center_x = (x1 + x2) // 2
                center_y = (y1 + y2) // 2
                cv2.circle(frame, (center_x, center_y), 5, (0, 0, 255), -1)
                
                frame_center_y = frame.shape[0] // 2
                
                if center_y < frame_center_y:  # Check if the line is above the center of the frame
                    print("Move upward")
                elif center_y >= frame_center_y:  # Check if the line is at or below the center of the frame
                    print("Move forward")

        cv2.imshow('Gate Line Detection', frame)
        if cv2.waitKey(30) & 0xFF == ord('q'):
            break

    cap.release()
    cv2.destroyAllWindows()

if __name__ == '__main__':
    video_path = '/mnt/z/Programação/Huginn Team/gatevideo.mp4'
    lower_red = np.array([0, 75, 75])
    upper_red = np.array([10, 255, 255])
    detect_gate(video_path, lower_red, upper_red)

File: gatedetection.cpp
#include <opencv2/opencv.hpp>

using namespace cv;

void detect_gate(const std::string& video_path, const cv::Scalar& lower_red, const cv::Scalar& upper_red)
{
    // Open the video file
    VideoCapture cap(video_path);

    // Check if the video file was opened successfully
    if (!cap.isOpened())
    {
        std::cout << "Error: Could not open video file." << std::endl;
        return;
    }

    // Loop over the frames in the video
    while (cap.isOpened())
    {
        // Read the next frame from the video
        Mat frame;
        cap >> frame;

        // Check if the frame was read successfully
        if (frame.empty())
        {
            break;
        }

        // Convert the frame to HSV color space
        Mat hsv_frame;
        cvtColor(frame, hsv_frame, COLOR_BGR2HSV);

        // Create a mask for the red color
        Mat mask;
        inRange(hsv_frame, lower_red, upper_red, mask);

        // Dilate and erode the mask to remove noise
        dilate(mask, mask, Mat(), Point(-1, -1), 1);
        erode(mask, mask, Mat(), Point(-1, -1), 1);

        // Detect the edges in the mask
        Mat edges;
        Canny(mask, edges, 50, 150);

        // Find the lines in the edges
        std::vector<Vec4i> lines;
        HoughLinesP(edges, lines, 1, CV_PI / 180, 50, 50, 5);

        // Draw the lines on the frame
        for (const Vec4i& line : lines)
        {
            cv::line(frame, Point(line[0], line[1]), Point(line[2], line[3]), Scalar(0, 255, 0), 2);
        }

        // Display the frame
        imshow("Gate Line Detection", frame);

        // Check if the user pressed the 'q' key to quit
        if (waitKey(30) & 0xFF == 'q')
        {
            break;
        }
    }

    // Close the video file
    cap.release();

    // Destroy all the windows
    destroyAllWindows();
}

int main(int argc, char** argv)
{
    // Get the path to the video file
    const std::string video_path = "/mnt/z/Programação/Huginn Team/gatevideo.mp4";

    // Get the lower and upper bounds of the red color
    const cv::Scalar lower_red = cv::Scalar(0, 75, 75);
    const cv::Scalar upper_red = cv::Scalar(10, 255, 255);

    // Detect the gate in the video
    detect_gate(video_path, lower_red, upper_red);

    return 0;
}